extension/sigv4authextension/extension.go

```
@@ -0,0 +1,207 @@
+// returns a new Sigv4Auth struct.
+func newSigv4Extension(cfg *Config, runtimeInfo string, logger *zap.Logger) (*Sigv4Auth, error) {
+	err := cfg.Validate()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Sigv4Auth{
+		cfg: cfg,
+		logger: logger,
+		runtimeInfo: runtimeInfo,
+	}, nil
+}
+
+// RoundTripperWithCredentials() 
+func RoundTripperWithCredentials(base http.RoundTripper, cfg *Config, creds *credentials.Credentials, runtimeInfo string) (http.RoundTripper, error) {
+	signer := sigv4.NewSigner(creds)
+
+	// Create the SigningRoundTripper struct
+	rt := SigningRoundTripper {
+		transport: base,
+		signer: signer,
+		region: cfg.Region,
+		service: cfg.Service,
+		runtimeInfo: runtimeInfo,
+	}
+	   
+	return &rt, nil
+}
+
+// getAWSConfig() is a helper function that gets the AWS Config
+// from the Config struct
+func getAWSConfig(cfg *Config) (*aws.Config, error) {
+	config := &aws.Config{Region: aws.String(cfg.Region)}
+	if cfg.STSEndpoint != "" {
+		if stsEndpoint, err := endpoints.GetSTSRegionalEndpoint(cfg.STSEndpoint); err == nil {
+			config.STSRegionalEndpoint = stsEndpoint
+		} else {
+			return nil, err
+		}
+	}
+	return config, nil
+}
+
+// getCredsFromConfig() is a helper function that gets AWS credentials
+// from the Config.
+func getCredsFromConfig(cfg *Config) (*credentials.Credentials, error) {
+	config, err := getAWSConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config: *config,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if cfg.RoleArn != "" {
+		// Get credentials from an assumeRole API call.
+		return stscreds.NewCredentials(sess, cfg.RoleArn, func(p *stscreds.AssumeRoleProvider) {
+			p.RoleSessionName = "otel-" + strconv.FormatInt(time.Now().Unix(), 10)
+		}), nil
+	}
+	// Get Credentials, either from ./aws or from environmental variables.
+	creds := sess.Config.Credentials
+	if creds == nil {
```

It looks like we should call `creds.Get()` once here to actually retrieve credentials, `creds` seems to just be a list of providers but is not actually validated

---

When I try to use the value from `creds.Get()`, I can't check it against nil anymore, nor does it return `*credentials.Credentials` but `credentials.Value` instead. Do we need to revisit the implementation of some of the methods/functions to use `credentials.Value` instead?

---

We don't need to store or use `creds.Get`, we just need to check the error, e.g. `_, err := creds.Get()` to fail fast on auth errors. The credential will be cached within the `creds` object until it expires so won't cause an extra lookup or anything like that

---

Added the check